Use single seed for randomization and add end-to-end test

jeannielynnmoulton opened the pull request:
This removes the use of the parameter `seed` that is passed around and instead puts everything into `numpy` `state`. 

This adds an end-to-end test based on `simple_example.py` that shows different seeds produce different results. It is admittedly not at all a unit test, but I wanted to get something running for these changes. 

NB: There is no test for the path that uses `Trachoma_Simulation` or `sim_Ind_MDA` yet, and this is slightly scary. I hacked together a way to get test_model.py at least running to show that these seed changes work, but it is not for committing yet.

The test is added to CI. I've modified the pull request triggers for CI because we should re-run tests when the PR is edited.

Other small changes include adding the qualifier `tf` for importing `trachoma_functions.py` as since I am unfamiliar with the code, this helped me better see the dependencies.

tlestang commented on .github/workflows/test_model.yaml:
```diff
@@ -26,3 +24,8 @@ jobs:
+      - name: Run test
+        run: |
+          cd tests
+          python -m unittest test_endtoend.py
```
I think we decided to use pytest to drive the tests?

jeannielynnmoulton replied:
Can do!

tlestang reviewed (commented):
Thanks @jeannielynnmoulton ! 

I didn't know about `numpy.random.{get, set}_state`. What is the benefit of capturing the state of the random generator compared to setting the seed before calling the model main funtion (`tf.run_single_simulation`). I looks to me like you then have to pass this state around down to every function., whereas setting the seed does so globally. Provided that nowhere in the called code the see is (re)set of course. But you removed both calls to `numpy.random.seed` anyway.

I remember from our last RSE meeting that we settled on pytest to write and drive automatic tests. In this case it would maybe simplify `test_endtoend.py` a bit since you wouldn't have to wrap both `test_` functions into a class. For simple cases like that I don't have a strong preference for pytest, but it's worth trying to keep it consistent over the models.

jeannielynnmoulton commented:
@tlestang I've changed this to you pytest now.

Regarding your comment about passing around the state - I can't see a way around this without massively changing the structure. When I don't pass the state and rely on the global setting, I don't get consistent results. I would like to create a Simulation class and a SimulationDriver that uses some sort of random state controller to get that under control.

Anyway, I know this is not beautiful, but it should allow us to progress I think.

I added a new commit at the end - the previous implementation use the same numpy state for every simulation - which obviously is not good :)

tlestang reviewed (approved):
Sorry, I should have approved this weeks ago.
